File: algorithm_introduction/ss26_maximum_flow/cpp/ford_fulkerson.cpp
#include <list>
#include <memory>
#include <set>
#include <stdio.h>
#include <string>
#include <utility>
#include <vector>

#define N 6

using namespace std;

struct Edge {
  int u;
  int v;
  int f;
  int c;
  int cf = c - f;
  string to_str() { return "(u: " + to_string(u) + ", v: " + to_string(v) + ", f: " + to_string(f) + ")"; }
};

vector<Edge> E[N]{};
vector<Edge *> _E[N]{};
Edge *E_m[][N];

void dfs_visit(vector<Edge *> _E[], int u, int t, int pth[]) {
  if (pth[t] >= 0 || u == t) {
    return;
  }
  for (auto e : _E[u]) {
    if (pth[e->v] < 0 && e->cf > 0) {
      pth[e->v] = u;
      dfs_visit(_E, e->v, t, pth);
    }
  }
}

vector<Edge *> dfs(vector<Edge *> _E[], int s, int t) {
  int pth[N];
  memset(pth, -1, N);

  dfs_visit(_E, s, t, pth);

  if (pth[t] < 0) {
    // Augmenting path is not found
    return vector<Edge *>{};
  } else {
    vector<Edge *> p;
    int cur = t;
    while (cur != s) {
      int pre = pth[cur];
      p.push_back(_E[pre][cur]);
      cur = pre;
    }
    return p;
  }
}

void ford_fulkerson(vector<Edge> E[], int s, int t) {
  vector<Edge> Ef[N];
  for (int i = 0; i < N; ++i) {
    for (auto e : E[i]) {
      Ef[e.v].emplace_back(e.v, e.u, e.c, e.f);
    }
  }
  for (int i = 0; i < N; ++i) {
    for (auto e : E[i]) {
      _E[i].push_back(&e);
      E_m[i][e.v] = &e;
    }
    for (auto e : Ef[i]) {
      _E[i].push_back(&e);
    }
  }
  while (true) {
    vector<Edge *> p = dfs(_E, s, t);
    if (p.empty()) {
      break;
    }
    int cf = p[0]->cf;
    for (auto e : p) {
      cf = min(cf, e->cf);
    }
    for (auto e : p) {
      if (E_m[e->u][e->v] != nullptr) {
        e->f += cf;
      } else {
        e->f -= cf;
      }
    }
  };
}

int main() {
  E[0] = vector<Edge>{Edge{0, 1, 0, 16}, Edge{0, 2, 0, 13}};
  E[1] = vector<Edge>{Edge{1, 3, 0, 12}};
  E[2] = vector<Edge>{Edge{2, 1, 0, 4}, Edge{2, 4, 0, 14}};
  E[3] = vector<Edge>{Edge{3, 5, 0, 20}};
  E[4] = vector<Edge>{Edge{4, 3, 0, 7}, Edge{4, 5, 0, 4}};

  ford_fulkerson(E, 0, 5);

  for (int i = 0; i < N; ++i) {
    for (auto e : E[i]) {
      printf("%s\n", e.to_str().c_str());
    }
  }
}

File: algorithm_introduction/ss26_maximum_flow/python/ford_fulkerson.py
from typing import List


N = 6


class Edge:
    def __init__(self, u, v, c, f):
        self.u = u
        self.v = v
        self.c = c
        self.f = f
        self.cf = c - f


def dfs_visit(_E: List[List[Edge]], u: int, t: int, pth: List[int]) -> None:
    if pth[t] >= 0 or u == t:
        return

    for e in _E[u]:
        if pth[e.v] < 0 and e.cf > 0:
            pth[e.v] = u
            dfs_visit(_E, e.v, t, pth)


def dfs(_E: List[List[Edge]], s: int, t: int) -> List[Edge]:
    pth = [-1 for i in range()]

    dfs_visit(_E, s, t, pth)

    if pth[t] < 0:
        # Augmenting path is not found
        return []
    else:
        p = []
        cur = t
        while cur != s:
            pre = pth[cur]
            p.append(_E[pre][cur])
            cur = pre
        return p


def ford_fulkerson(E: List[List[Edge]], s: int, t: int) -> None:
    global N

    Ef = [[] for _ in range(N)]
    for e in E:
        Ef[e.v].append(Edge(e.v, e.u, e.f, e.c))

    _E = [[None for _ in range(N)] for _ in range(N)]
    E_m = [[None for _ in range(N)] for _ in range(N)]
    for e in E:
        _E[e.u][e.v] = e
        E_m[e.u][e.v] = e
    for e in _E:
        _E[e.u][e.v] = e

    while True:
        p = dfs(_E, s, t)

        if len(p) == 0:
            break

        min_cf = min(p, key=lambda x: x.cf)
        for e in p:
            if _E[e.u][e.v] is not None:
                e.f += min_cf
            else:
                e.f -= min_cf


E = [
    [Edge(0, 1, 0, 16), Edge(0, 2, 0, 13)],
    [Edge(1, 3, 0, 12)],
    [Edge(2, 1, 0, 4), Edge(2, 4, 0, 14)],
    [Edge(3, 5, 0, 20)],
    [Edge(4, 3, 0, 7), Edge(4, 5, 0, 4)]
]

ford_fulkerson(E, 0, 5)
